File: typegraph/python/typegraph/runtimes/python.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import ast
import hashlib
import inspect
from typing import List, Optional, Tuple

from astunparse import unparse
from attr import field
from attrs import frozen

from typegraph import effects
from typegraph.effects import Effect
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.utils.attrs import always
from typegraph.graph.nodes import Node
from typegraph.graph.builder import Collector


class DefinitionCollector(ast.NodeTransformer):
    @classmethod
    def collect(cls, function):
        source = inspect.getsource(function).lstrip()
        tree = ast.parse(source)
        ret = cls()
        ret.visit(tree)
        return ret.lambdas, ret.defs

    def __init__(self):
        super().__init__()
        self.lambdas = []
        self.defs = []

    def visit_Lambda(self, node):
        self.lambdas.append(unparse(node).strip())

    def visit_FunctionDef(self, node):
        self.defs.append((node.name, unparse(node).strip()))


@frozen
class Python(Runtime):
    """
    [Documentation](https://metatype.dev/docs/reference/runtimes/python)
    """

    runtime_name: str = always("python_wasi")

    def from_lambda(self, function):
        lambdas, _defs = DefinitionCollector.collect(function)
        assert len(lambdas) == 1
        fn = str(lambdas[0])
        m = hashlib.sha256()
        m.update(fn.encode("utf-8"))
        return LambdaMat(self, m.hexdigest(), fn)

    def from_def(self, function):
        _lambdas, defs = DefinitionCollector.collect(function)
        assert len(defs) == 1
        name, fn = defs[0]
        return DefMat(self, name, fn)


@frozen
class LambdaMat(Materializer):
    runtime: Runtime
    name: str
    fn: str
    materializer_name: str = always("lambda")
    effect: Effect = field(kw_only=True, default=effects.none())


@frozen
class DefMat(Materializer):
    runtime: Runtime
    name: str
    fn: str
    materializer_name: str = always("def")
    effect: Effect = field(kw_only=True, default=effects.none())


# Import function from a module
@frozen
class ImportFunMat(Materializer):
    mod: "PyModuleMat" = field()
    name: str = field(default="main")
    secrets: Tuple[str] = field(kw_only=True, factory=tuple)
    effect: Effect = field(kw_only=True, default=effects.none())
    runtime: Python = field(
        kw_only=True, factory=Python
    )  # should be the same runtime as `mod`'s
    materializer_name: str = always("import_function")
    collector_target = always(Collector.materializers)

    @property
    def edges(self) -> List[Node]:
        return super().edges + [self.mod]

    def data(self, collector: Collector) -> dict:
        data = super().data(collector)
        data["data"]["mod"] = collector.index(self.mod)
        return data


@frozen
class PyModuleMat(Materializer):
    file: Optional[str] = field(default=None)
    secrets: Tuple[str] = field(kw_only=True, factory=tuple)
    code: Optional[str] = field(kw_only=True, default=None)
    runtime: Python = field(kw_only=True, factory=Python)
    materializer_name: str = always("pymodule")
    effect: Effect = always(effects.none())

    def __attrs_post_init__(self):
        if self.file is None:
            raise Exception("you must provide the source file")
        object.__setattr__(self, "code", f"file:{self.file}")

    def imp(
        self, name: str = "main", *, effect: Effect = effects.none(), **kwargs
    ) -> ImportFunMat:
        return ImportFunMat(
            self,
            name,
            runtime=self.runtime,
            secrets=self.secrets,
            effect=effect,
            **kwargs,
        )

File: typegraph/deno/src/runtimes/python.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Effect } from "../gen/interfaces/metatype-typegraph-runtimes.d.ts";
import { Materializer, Runtime } from "./mod.ts";

interface LambdaMat extends Materializer {
  fn: string;
  effect: Effect;
}

interface DefMat extends Materializer {
  fn: string;
  name: string;
  effect: Effect;
}

interface PythonImport {
  name: string;
  module: string;
  secrets?: Array<string>;
  effect?: Effect;
}

interface ImportMat extends Materializer {
  module: string;
  name: string;
}

export class PythonRuntime extends Runtime {
  constructor() {
    super(runtimes.registerPythonRuntime());
  }
  fromLambda<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(inp: I, out: O, { code }: { code: string }) {
    const matId = runtimes.fromPythonLambda(
      {
        runtime: this._id,
        effect: { tag: "none" },
      },
      {
        fn: code, // not formatted
        runtime: this._id,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      fn: code,
    } as LambdaMat);
  }

  fromDef<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(inp: I, out: O, { code }: { code: string }) {
    const name = code.trim().match(/def\s+([A-Za-z0-9_]+)/)?.[1];
    if (name == undefined) {
      throw new Error(`unable to extract def name from source code ${code}`);
    }
    const matId = runtimes.fromPythonDef(
      {
        runtime: this._id,
        effect: { tag: "none" },
      },
      {
        name: name,
        fn: code,
        runtime: this._id,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      name,
      fn: code,
    } as DefMat);
  }

  import<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    { name, module, effect = { tag: "none" }, secrets = [] }: PythonImport,
  ): t.Func<P, I, O, ImportMat> {
    const base = {
      runtime: this._id,
      effect,
    };

    const matId = runtimes.fromPythonModule(base, {
      file: module,
      runtime: this._id,
    });

    const pyModMatId = runtimes.fromPythonImport(base, {
      module: matId,
      funcName: name,
      secrets,
    });

    return t.func(inp, out, {
      _id: pyModMatId,
      module,
      name,
    });
  }
}
